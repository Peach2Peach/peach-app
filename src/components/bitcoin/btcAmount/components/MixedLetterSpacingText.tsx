import { TextStyle } from 'react-native'
import { Text } from '../../..'
import { BTCAmountChar } from './BTCAmountChar'
import { getLetterSpacing, getNewNumber } from './helpers'

type Props = {
  value: number
  style: (false | TextStyle)[]
  white: boolean
}

export const MixedLetterSpacingText = ({ value, style, white }: Props) => {
  const newNum = getNewNumber(value)
  const { dot, whiteSpace, digit } = getLetterSpacing(style)

  return (
    <Text style={style}>
      {newNum.map((char, index) => {
        const reduceOpacity
          = index < newNum.findIndex((c) => c !== '0' && Number(c) > 0) || (value === 0 && index !== newNum.length - 1)
        const letterSpacing = char === '.' ? dot : char === ' ' ? whiteSpace : digit
        const props = { style, reduceOpacity, white, char, letterSpacing }
        return <BTCAmountChar key={index} {...props} />
      })}
    </Text>
  )
}
