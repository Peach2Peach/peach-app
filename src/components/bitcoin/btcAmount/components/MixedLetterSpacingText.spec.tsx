import { createRenderer } from 'react-test-renderer/shallow'
import { MixedLetterSpacingText } from './MixedLetterSpacingText'

describe('MixedLetterSpacingText', () => {
  const value = 1280000
  const renderer = createRenderer()
  it('should render correctly', () => {
    renderer.render(<MixedLetterSpacingText value={value} style={[]} white={false} />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
  it('should render correctly with white', () => {
    renderer.render(<MixedLetterSpacingText value={value} style={[]} white />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
  it('should render correctly if the value is 0', () => {
    renderer.render(<MixedLetterSpacingText value={0} style={[]} white={false} />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})
