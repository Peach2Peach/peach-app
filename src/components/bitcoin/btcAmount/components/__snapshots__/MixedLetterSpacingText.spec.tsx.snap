// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MixedLetterSpacingText should render correctly 1`] = `
<PeachText
  style={[]}
>
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="."
    isError={false}
    letterSpacing={-4.62}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="1"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char=" "
    isError={false}
    letterSpacing={-7.699999999999999}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="2"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="8"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char=" "
    isError={false}
    letterSpacing={-7.699999999999999}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
</PeachText>
`;

exports[`MixedLetterSpacingText should render correctly if the value is 0 1`] = `
<PeachText
  style={[]}
>
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="."
    isError={false}
    letterSpacing={-4.62}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char=" "
    isError={false}
    letterSpacing={-7.699999999999999}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char=" "
    isError={false}
    letterSpacing={-7.699999999999999}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={false}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
</PeachText>
`;

exports[`MixedLetterSpacingText should render correctly with error 1`] = `
<PeachText
  style={[]}
>
  <BTCAmountChar
    char="0"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="."
    isError={true}
    letterSpacing={-4.62}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={true}
    style={[]}
  />
  <BTCAmountChar
    char="1"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char=" "
    isError={true}
    letterSpacing={-7.699999999999999}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="2"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="8"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char=" "
    isError={true}
    letterSpacing={-7.699999999999999}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
  <BTCAmountChar
    char="0"
    isError={true}
    letterSpacing={-1.65}
    reduceOpacity={false}
    style={[]}
  />
</PeachText>
`;
